### PR TITLE
Don't crash on a null host

### PR DESCRIPTION
**Pull request by swankjesse**
Closes: https://github.com/square/okhttp/issues/5770

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/internal/connection/RouteSelector.kt [hunk 1]
```diff
@@ -95,17 +95,19 @@ class RouteSelector(
...
-      } else {
-        immutableListOf(Proxy.NO_PROXY)
-      }
+      val proxiesOrNull = address.proxySelector.select(uri)
+      if (proxiesOrNull.isNullOrEmpty()) return@run immutableListOf(Proxy.NO_PROXY)
+
+      return@run proxiesOrNull.toImmutableList()
     }
     nextProxyIndex = 0
     eventListener.proxySelectEnd(call, url, proxies)
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/connection/RouteSelector.kt]: alternately:
```
 private fun resetNextProxy(url: HttpUrl, proxy: Proxy?) {
    eventListener.proxySelectStart(call, url)
    proxies = if (proxy != null) {
      listOf(proxy)
    } else {
      val uri = url.toUri()
      if (uri.host == null) {
        immutableListOf(Proxy.NO_PROXY)
      } else {
        val proxiesOrNull = address.proxySelector.select(uri)
        if (proxiesOrNull.isNullOrEmpty()) {
          immutableListOf(Proxy.NO_PROXY)
        } else {
          proxiesOrNull.toImmutableList()
        }
      }
    }
    nextProxyIndex = 0
    eventListener.proxySelectEnd(call, url, proxies)
  }
```

File: okhttp/src/main/java/okhttp3/internal/connection/RouteSelector.kt [hunk 2]
```diff
@@ -95,17 +95,19 @@ class RouteSelector(
   /** Prepares the proxy servers to try. */
   private fun resetNextProxy(url: HttpUrl, proxy: Proxy?) {
     eventListener.proxySelectStart(call, url)
-    proxies = if (proxy != null) {
+    proxies = run {
```
Comment (yschimke) [inline okhttp/src/main/java/okhttp3/internal/connection/RouteSelector.kt]: is it cleaner just to have an inline function than to use run and return@run?

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/internal/connection/RouteSelector.kt]: Perhaps? I changed to use a local function.

**Comment by swankjesse** [comment]
Completely agree that we’ve fallen into a bizarre invalid-fest!

In this particular code we merely need to stay alive long enough for the DNS request to return 0 results. Without this the default proxy selector crashes with an IllegalArgumentException, which is really bad for tools like crawlers that could encounter such URLs in the wild.

I don’t expect this particular modeling failure to push itself deeper into our code. Although HttpUrl is too lenient, it’s self-consistent and the only places we get hurt is when we take sketchy URLs like `https://>/` and integrate them into other systems: DNS, proxy selection, cookie stores, etc.

The other idea I had for `https://>/` was to replace its impossible hostname with a well-formed but unresolvable URL, Like `https://-/` or `https://0/`. Ultimately those also feel bad!

File: okhttp/src/test/java/okhttp3/HttpUrlTest.java [hunk 3]
```diff
@@ -440,6 +440,13 @@ HttpUrl parse(String url) {
         .test(Component.HOST);
   }
 
+  /** This one's ugly: the HttpUrl's host is non-empty, but the URI's host is null. */
+  @Test public void hostContainsOnlyStrippedCharacters() throws Exception {
+    HttpUrl url = parse("http://>/");
+    assertThat(url.host()).isEqualTo(">");
```
Comment (monkey-mas) [inline okhttp/src/test/java/okhttp3/HttpUrlTest.java]: Should we add validation in `HttpUrl` that prevents the user from having an _invalid_ hostname?

Comment (swankjesse) [reply okhttp/src/test/java/okhttp3/HttpUrlTest.java]: Yes, but...

We’d need to define what a valid hostname is. That’s difficult, and lots of libraries that have attempted this have failed.

**Pull request merged**